### ISSUE TITLE
Autoshutdown not working on a Koob Touch

### ISSUE THREAD

**Issue by MrPick1es** [koreader/koreader#8039]
Hi there,
as the title states I can't seem to get auto shutdown to work. Auto suspend on the other hand does work.

I'm using Kobo Touch (2gb). I have auto suspend set to 15min and auto shutdown set to 1 hr. I've tried a few different times for the auto shutdown, such as 0.5 hr, 0.2 hr and it wouldn't shut down itself. I understand it's a pretty minor issue but I figured I would post about it in case nobody was aware of the problem. Thanks and let me know if there is anything I can do to help!

**Comment by NiLuJe** [comment]
Cross ref to the recent autoshutdown issue on a Glo for more details.

That said, I'm not even sure it's supposed to work on a *sane* Touch to begin with, but a Touch these days? I wouldn't hold my breath ;).

**Comment by MrPick1es** [comment]
Thanks for the comment I will look into that thread. I realize it's an old device, it was given to me by someone who had it just lying around and I figured I would give it a try.

**Comment by djibux** [comment]
I am experiencing the issue on a Kobo Glo.  I kept having no battery left until I figured out the Kobo never shuts down.

**Comment by NiLuJe** [comment]
IIRC, the Glo stuff was related to a dying rtc battery or something (or at least erratic ioctl returns), and the user just patched those to pass.

**Comment by yfede** [comment]
I have the same issue with autoshutdown not working on my Kobo Aura (1st edition, model N514).
I'm currently on release 2022.08 (the latest at the time of writing)

Also, testing this has become problematic since when autoshutdown timeout can only be expressed in days and hours, with a minimum of 1 day. IMHO it would make sense to also allow timeouts of less than a day. Maybe I'm missing something in how the input is entered.

**Comment by NiLuJe** [comment]
Yeah, you won't be able to set a timer below 24H in the UI, you'll have to update the settings manually (*without* KOReader running. USBMS would work, for instance).

Do note that if you're in the habit of plugging/unplugging a sleeping device, an issues around that was fixed in current nightlies, a few days after the 2022.08 release.

If all else fails, as usual, we're going to need verbose debug logs to see what's happening ([Tools]> More tools > Developer options).

**Comment by yfede** [comment]
Thanks, I have set it to 10 minutes editing the file, and I'm now waiting for the time to pass and see what happens.
I haven't found "Developer options" under "More tools", perhaps that's only in the nightly builds?

**Comment by yfede** [comment]
I understood the nature of the problem.

**TL;DR:** the autoshutdown check is triggered by a timer, but said timer seem to have a countdown variable limited to 16 bits, thus any timeout value greater than 65536 seconds (18h12m) causes the timer to fire early and autoshutdown doesn't happen because KOReader realizes that it's not time to autoshutdown yet.

I have done some testing, and there seem to be an overflow error somewhere that's causing this issue.

### 10 minutes test:
First I have tried setting a autoshutdown timeout of **600 seconds** (10 minutes) and the Kobo did shut down correctly after that time. Here is the corresponding crash log: [crash 600 success.log](https://github.com/koreader/koreader/files/9684021/crash.600.success.log)
Brief excerpt: 
at **11:21** I put it to sleep:
```
09/29/22-11:21:36 INFO  Kobo suspend: asking for a suspend to RAM . . .
```
and 10 minutes later (at **11:31**) an internal countdown seem to have terminated:
```
09/29/22-11:31:32 DEBUG WakeupMgr:isWakeupAlarmScheduled: An alarm is scheduled for 1664443877 (2022-09-29 11:31:17 +0200)
validateWakeupAlarmByProximity:	
task              @ 1664443877 (2022-09-29 11:31:17 +0200)	
last set alarm    @ 1664443877 (2022-09-29 11:31:17 +0200)	
current rtc alarm @ 1664443877 (2022-09-29 11:31:17 +0200)	
current time is     1664443892 (2022-09-29 11:31:32 +0200)
```
The "task" here must be the autoshutdown. The "current time" has reached the "task" time (just a few seconds past it), and the Kobo realizes that it's time to shut itself down.

### 24 hours test:
Some time later I tried setting an autoshutdown timeout of 86400 (24 hours) instead, which I know doesn't work, and put the Kobo again to sleep (at **12:15**):

    09/29/22-12:15:41 INFO  Kobo suspend: asking for a suspend to RAM . . .

I expected the wakeup event to happen only after 24 hours, meaning the next day around the same time. But just around 6 hours later (at **18:03** the same day):

    09/29/22-18:03:21 DEBUG WakeupMgr:isWakeupAlarmScheduled: An alarm is scheduled for 1664532923 (2022-09-30 12:15:23 +0200)
    validateWakeupAlarmByProximity:	
    task              @ 1664532923 (2022-09-30 12:15:23 +0200)	
    last set alarm    @ 1664532923 (2022-09-30 12:15:23 +0200)	
    current rtc alarm @ 1664532923 (2022-09-30 12:15:23 +0200)	
    current time is     1664467401 (2022-09-29 18:03:21 +0200)

And this time case the "current time" is several hours early than the "task" time, in fact it's still on the 2022-09-29 while the task time is on the next day 2022-09-30. 
Seeing that the task time isn't arrived yet, the Kobo does nothing and doesn't shut itself down.
No more wakeup events happen in the next 24 hours, meaning that the wakeup event that happened just 6 hours after the Kobo was put to sleep, was instead intended to be the one happening 24 hours after sleep time, to shut the Kobo off.
(full log here: [crash 86400 fail.log](https://github.com/koreader/koreader/files/9684297/crash.86400.fail.log))

### So, why the early wakeup?

I believe the wakeup is triggered by a timer with a 16 bit countdown.
Trying to set that countdown to 86400, actually overflows those 16 bits, which can only store values up to 65536, and the value that ends being written there is just the excess value 86400 - 65536 = 20864 seconds.

That value corresponds to about 5 hours and 48 minutes, which is exactly the time elapsed between 12:15 and 18:03, when the timer fired early.

### Final check: causing the timer to fire early on purpose
To check my findings, today I have tried setting a timeout of 66436 seconds, which should correspond to around 18h 27m, but if I got this right, that should cause the timer to fire much earlier, in fact I expected the timer to fire exactly 66436 - 65536 = 900 seconds (15 minutes) after putting the Kobo to sleep.
And in fact I put it to sleep at **13:04**:

    09/30/22-13:04:52 INFO  Kobo suspend: asking for a suspend to RAM . . .

and 15 minutes later, at **13:19** I see:

    09/30/22-13:19:48 DEBUG WakeupMgr:isWakeupAlarmScheduled: An alarm is scheduled for 1664602309 (2022-10-01 07:31:49 +0200)
    validateWakeupAlarmByProximity:	
    task              @ 1664602309 (2022-10-01 07:31:49 +0200)	
    last set alarm    @ 1664602309 (2022-10-01 07:31:49 +0200)	
    current rtc alarm @ 1664602309 (2022-10-01 07:31:49 +0200)	
    current time is     1664536788 (2022-09-30 13:19:48 +0200)

(full log here: [crash 66436.log](https://github.com/koreader/koreader/files/9684308/crash.66436.log))

### Possible solutions

The simplest solution would be to increase the size of the countdown variable to 32 bits.
If that's not possible due to hardware limitations, I guess the only other sensible idea is to handle timeouts longer than 65536 as a sequence of wakeup events at intervals close to the max allowed value (for example, 64000 seconds), until one of those events finally realizes that the autoshutdown time is less than 65536 seconds away, and sets the next (and final) wakeup event with the remaining countdown.

**Comment by Frenzie** [comment]
That's amazingly detailed. 😍

I don't think there should be hardware limitations with 32-bit. @zwim redid it I think since I wrote the initial implementation?

**Comment by NiLuJe** [comment]
That still looks like a misbehaving RTC to me ;o).

As @Frenzie said, we're in Lua-land, everything is a double, so we have 52 bits of precision available for anything *we* do.

Enter the RTC. Some older RTCs, do, in fact, have limitations (usually, for wakeups more than 24H in the future).

What does `/proc/driver/rtc` say on your device?

**Comment by NiLuJe** [comment]
There's not really anything else that could cause a wakeup without any input events going through, but it could be interesting to see what the kernel log has to say about this (run `klogd` once so it gets sent to the syslog & timestamped, then check `logread`).

**Comment by yfede** [comment]
> What does `/proc/driver/rtc` say on your device?

```
rtc_time	: 15:36:31
rtc_date	: 2022-09-30
alrm_time	: 13:43:24
alrm_date	: 2022-10-01
alarm_IRQ	: yes
alrm_pending	: no
24hr		: yes

```

**Comment by NiLuJe** [comment]
Okay, that looks sound (which makes sense, the ioctl would have failed otherwise; and the alarm time is okay, too).

**Comment by yfede** [comment]
> run `klogd` once so it gets sent to the syslog & timestamped, then check `logread`

Ok, to be clear about what I've done:

I have set autoshutdown timeout to 65716 (= 65536 + 180) so that the misbehaving timer fires 180s (3 minutes) after putting it to sleep.

I have then
1. run klogd
2. waited some time
3. put the kobo to sleep at 17:57:38
4. waited more than the expected 3 minutes (I did 5)
5. launched logread to see what it had to say between 17:57:38 and say 18:02:00.

<details>
<summary>Here is the relevant part. notice the expected 3 minutes gap happening between 17:57:41 and 18:00:40</summary>

```
Sep 30 17:57:39 kernel: Waiting for arm_podf changed!
Sep 30 17:57:39 kernel: current wp=0
Sep 30 17:57:39 kernel: [msp430_setalarm-798] alarm 65716
Sep 30 17:57:39 kernel: ntx_get_battery_vol : full !! 4200000
Sep 30 17:57:39 kernel: ntx_get_battery_vol : full !! 4200000
Sep 30 17:57:39 kernel: ntx_get_battery_vol : full !! 4200000
Sep 30 17:57:39 kernel: ntx_get_battery_vol : full !! 4200000
Sep 30 17:57:39 kernel: ntx_get_battery_vol : full !! 4200000
Sep 30 17:57:39 kernel: FL PWM off command
Sep 30 17:57:39 kernel: TPS65185 temprature data = 0x17,23
Sep 30 17:57:39 kernel: Waiting for arm_podf changed!
Sep 30 17:57:39 kernel: current wp=1
Sep 30 17:57:39 kernel: Waiting for arm_podf changed!
Sep 30 17:57:39 kernel: current wp=2
Sep 30 17:57:40 kernel: Waiting for arm_podf changed!
Sep 30 17:57:40 kernel: current wp=0
Sep 30 17:57:40 dhcpcd[1292]: sending signal ARLM to pid 1234
Sep 30 17:57:40 dhcpcd[1234]: received signal ALRM from PID 1292, releasing
Sep 30 17:57:40 dhcpcd[1234]: eth0: removing interface
Sep 30 17:57:40 dhcpcd[1234]: eth0: releasing lease of 192.168.1.245
Sep 30 17:57:40 dhcpcd[1234]: eth0: sending RELEASE with xid 0xfedf866
Sep 30 17:57:40 dhcpcd[1234]: eth0: deleting IP address 192.168.1.245/24
Sep 30 17:57:40 dhcpcd[1234]: eth0: deleting route to 192.168.1.0/24
Sep 30 17:57:40 dhcpcd[1234]: eth0: deleting default route via 192.168.1.254
Sep 30 17:57:40 dhcpcd[1234]: eth0: executing `/libexec/dhcpcd-run-hooks' STOP
Sep 30 17:57:40 dhcpcd[1292]: waiting for pid 1234 to exit
Sep 30 17:57:40 dhcpcd[1234]: eth0: executing `/libexec/dhcpcd-run-hooks' STOPPED
Sep 30 17:57:40 kernel: [FL_off_func-1511]FL PWR off
Sep 30 17:57:40 dhcpcd[1234]: exited
Sep 30 17:57:40 dhcpcd[1318]: dhcpcd not running
Sep 30 17:57:40 dhcpcd[1331]: dhcpcd not running
Sep 30 17:57:40 wpa_supplicant[1195]: ioctl[SIOCSIWESSID]: Resource temporarily unavailable
Sep 30 17:57:40 wpa_supplicant[1195]: eth0: CTRL-EVENT-DISCONNECTED bssid=20:b0:01:da:c4:29 reason=3 locally_generated=1
Sep 30 17:57:40 wpa_supplicant[1195]: ioctl[SIOCSIWESSID]: Resource temporarily unavailable
Sep 30 17:57:40 wpa_supplicant[1195]: eth0: CTRL-EVENT-TERMINATING 
Sep 30 17:57:40 kernel: Waiting for arm_podf changed!
Sep 30 17:57:40 kernel: current wp=1
Sep 30 17:57:41 kernel: Waiting for arm_podf changed!
Sep 30 17:57:41 kernel: current wp=2
Sep 30 17:57:41 kernel: Waiting for arm_podf changed!
Sep 30 17:57:41 kernel: DVFS is stopped
Sep 30 17:57:41 kernel: Wifi / BT power control 0 rst 113
Sep 30 17:57:41 kernel: cd_status=1 removed
Sep 30 17:57:41 kernel: cd_status=1 removed
Sep 30 17:57:41 kernel: mmc1: Card removed and resetting controller.
Sep 30 17:57:41 kernel: mmc1: card 0001 removed
Sep 30 18:00:40 kernel: PM: Preparing system for mem sleep
Sep 30 18:00:40 kernel: Freezing user space processes ... freeze 1!!!!!!!!!!!!!!!!!!!!
Sep 30 18:00:40 kernel: (elapsed 0.01 seconds) done.
Sep 30 18:00:40 kernel: Freezing remaining freezable tasks ... (elapsed 0.01 seconds) done.
Sep 30 18:00:40 kernel: PM: Entering mem sleep
Sep 30 18:00:40 kernel: Suspending console(s) (use no_console_suspend to debug)
Sep 30 18:00:40 kernel: mma7660 current mode=0x41->0x40
Sep 30 18:00:40 kernel: mma7660_irqwake_enable(0);
Sep 30 18:00:40 kernel: mma7660_irq_enable(0);
Sep 30 18:00:40 kernel: MMA7660 suspended
Sep 30 18:00:40 kernel: USB Host suspend begins
Sep 30 18:00:40 kernel: ehci_fsl_drv_suspend, pm event 
Sep 30 18:00:40 kernel: [sdhci_suspend-1838] skip suspend for ESD (mmc0)
Sep 30 18:00:40 kernel: PM: suspend of devices complete after 4.452 msecs
Sep 30 18:00:40 kernel: suspend wp cpu=1000000000
Sep 30 18:00:40 kernel: PM: late suspend of devices complete after 0.186 msecs
Sep 30 18:00:40 kernel: PM: early resume of devices complete after 0.120 msecs
Sep 30 18:00:40 kernel: [msp_int-1883] MSP430 interrupt triggered !!!
Sep 30 18:00:40 kernel: [elan_touch_report_data-349] undefined packet 0xFF.
Sep 30 18:00:40 kernel: 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 
Sep 30 18:00:40 kernel: [sdhci_resume-1901] skip resume for ESD (mmc0)
Sep 30 18:00:40 kernel: ehci fsl drv resume begins: Host 1
Sep 30 18:00:40 kernel: ehci_fsl_drv_resume,pm event, wait for wakeup irq if needed
Sep 30 18:00:40 kernel: ehci_fsl_bus_resume, Host 1
Sep 30 18:00:40 kernel: mma7660_irq_enable(1);
Sep 30 18:00:40 kernel: mma7660 sensor triggered ...
Sep 30 18:00:40 kernel: mma7660 sensor triggered ...
Sep 30 18:00:40 kernel: mma7660_irqwake_enable(1);
Sep 30 18:00:40 kernel: MMA7660 resumed
Sep 30 18:00:40 kernel: PM: resume of devices complete after 336.723 msecs
Sep 30 18:00:40 kernel: PM: Finishing wakeup.
Sep 30 18:00:40 kernel: Restarting tasks ... done.
Sep 30 18:00:42 kernel: ehci_fsl_bus_suspend, Host 1
Sep 30 18:00:55 kernel: read_alarm: 1 122/9/1 10:12:55
```
</details>

**Comment by yfede** [comment]
Side note: I'm by no mean an expert in this kind of logs, but it seems to me that those messages between
`Sep 30 18:00:40 kernel: PM: Preparing system for mem sleep`
and
`Sep 30 18:00:40 kernel: PM: late suspend of devices complete after 0.186 msecs`
belong to the suspending phase and have been put in the queue just before suspend, and appear with a timestamp after 18:00 only because they have been flushed to the log after resume.

I guess the real first message in the log belonging to the resume phase is:
`Sep 30 18:00:40 kernel: PM: early resume of devices complete after 0.120 msecs`

**Comment by NiLuJe** [comment]
Yup, that's perfectly normal :).

**Comment by NiLuJe** [comment]
Oohkay, that's... interesting.

I can reproduce the apparent overflow on a H2O, let me check on what I actually tested this on instead (i.e., more recent, less crappy devices).

(It definitely appears to be the RTC doing wonky shit, though, so, might be extremely annoying to workaround...).

**Comment by NiLuJe** [comment]
Yeah, can't reproduce on a Forma, so at least I wasn't entirely mad when I wrote that ;o).

**Comment by NiLuJe** [comment]
(I would wager it's an issue that's only found on i.MX 50x devices?)

**Comment by NiLuJe** [comment]
Hopefully the commit above should do the trick?

**Comment by NiLuJe** [comment]
Could you confirm what `/sys/class/rtc/rtc0/name` is set to on your device?

**Comment by yfede** [comment]
Sure, it says:

```
# cat /sys/class/rtc/rtc0/name

pmic_rtc
```

**Comment by yfede** [comment]
I don't speak LUA but I think I can grasp the idea behind your approach. 

Correct me if I'm wrong: in case the next desired wakeup is more than 0xFFFF seconds in the future, you add to the schedule some "dummy" wakeup events, which happen 0xFFFF seconds apart and that do nothing, but serve the purpose of making sure that the RTC never has to count to a number of seconds greater than that?

**Comment by NiLuJe** [comment]
> Correct me if I'm wrong: in case the desired wakeup is more than 0xFFFF seconds in the future, you add to the schedule some "dummy" wakeup events before it, which happen 0xFFFF seconds apart and that do nothing, but serve the purpose of making sure that the RTC never has to count to a number of seconds greater than UINT16_MAX between them?

Yup, that's the idea!

When WakeupMgr pops a task from the list, it automatically schedules the alarm of the next one (if any), so this ensures the chain isn't broken without having to resort to uglier nastiness (like making the "jump" callbacks the thing that would set the alarm)).
(The fact that the RTC speaks in absolute time helps, here).

**Comment by NiLuJe** [comment]
Good news: the H2O had shutdown after its two hops!

(And I caught a bug that prevented the screensaver to show on the autoshutdown shutdown in the process, yay).

Should make it to tomorrow's nightly ;).

Many thanks for your extremely through investigation, because I'm pretty sure this was reported a couple times before, and I just went: "meh' probably a messed up rtc" (which... it technically is, but, still ;p).

**Comment by yfede** [comment]
Sorry for having disappeared. You're welcome! 
I'm a developer myself and I know that a well written, reproducible bug report does a great deal in speeding up the resolving process ;)

**Comment by yfede** [comment]
Ah something I forgot to ask before: do I need to do something in particular to undo the launch of `kmond` ? 

I'd like to bring my Kobo back to its pre-debugging condition, and I'm unsure whether `kmond` now runs every time I open KOReader, unnecessarily using resources. 

Thanks

**Comment by NiLuJe** [comment]
You meant `klogd`?

You can just kill it, or reboot your device (it's a standard *nix system utility, provided by busybox here).

The only KOReader-specific thing you might have done was enable verbose debug logging, so you can just disable that (and restart KOReader) ;).
